sshd-core/src/main/java/org/apache/sshd/common/future/DefaultSshFuture.java

```diff
@@ -106,10 +106,9 @@ public void setValue(Object newValue) {
             }
 
             result = (newValue != null) ? newValue : GenericUtils.NULL;
+            notifyListeners();
```

Comment:
Notifying the listener while under lock is extremely dangerous. I also fail to see the problem and why this solves it...

Reply:
The problem happens after the issue SSHD-848 was fixed. When a client session is closed listeners are added when the OpenFuture is done. If the listeners are NOT notified, the local port forwarding channel is closed whilst bytes are already flowing from client to the target server. This causes some bytes to be lost in transit. In the current code notifyListener() is called after a the synchronized block. Only with our code change we fixed the problem.
Do you understand what I mean ?

Reply:
In order to better clarify the issue, what happens is that during execution of method `DefaultSshFuture.setValue`, after the synchronized block is executed _but before or during_ `notifyListeners`  is called, the method `addListener` is called (see `DefaultForwardingFilter.sessionClosed`) by another thread so that `notifyNow` is `true` and the listener is executed immediately
`
...
        if (notifyNow) {
            notifyListener(listener);
        }
...
`
Not respecting the ordering the listeners are added. In our case

`                        openFuture.addListener(f -> {
                            channel.close(false);        
                        });
`
is executed while the existing listeners are called inside `setValue` so that all packets after `channel.close(false)` are not sent